presto-main/src/main/java/com/facebook/presto/operator/BlockEncodingBuffers.java

```diff
@@ -0,0 +1,363 @@
+    private boolean hasEncodedNulls;
+
+    public static BlockEncodingBuffers createBlockEncodingBuffers(DecodedBlockNode decodedBlockNode)
+    {
+        requireNonNull(decodedBlockNode, "decodedBlockNode is null");
+
+        Object decodedBlock = decodedBlockNode.getDecodedBlock();
+
+        // Skip the Dictionary/Rle block node. The mapping info is not needed when creating buffers.
+        if (decodedBlock instanceof DictionaryBlock) {
+            decodedBlockNode = decodedBlockNode.getChildren().get(0);
+            decodedBlock = decodedBlockNode.getDecodedBlock();
+        }
+        else if (decodedBlock instanceof RunLengthEncodedBlock) {
+            decodedBlockNode = decodedBlockNode.getChildren().get(0);
+            decodedBlock = decodedBlockNode.getDecodedBlock();
+        }
+
+        verify(!(decodedBlock instanceof DictionaryBlock), "Nested RLEs and dictionaries are not supported");
+        verify(!(decodedBlock instanceof RunLengthEncodedBlock), "Nested RLEs and dictionaries are not supported");
+
+        if (decodedBlock instanceof LongArrayBlock) {
+            return new LongArrayBlockEncodingBuffers();
+        }
+
+        throw new IllegalArgumentException("Unsupported encoding: " + decodedBlock.getClass().getSimpleName());
+    }
+
+    public abstract void resetBuffers();
+
+    public void setNextBatch(int positionsOffset, int batchSize)
+    {
+        this.positionsOffset = positionsOffset;
+        this.batchSize = batchSize;
+    }
+
+    public abstract void copyValues();
+
+    public abstract void serializeTo(SliceOutput output);
+
+    public abstract long getRetainedSizeInBytes();
+
+    public abstract long getSerializedSizeInBytes();
+
+    public abstract void accumulateRowSizes(int[] rowSizes);
+
+    // positionOffsets: the indexes to the positions in current buffer, which point to the boundaries of top level rows.
+    // positionCount: top level position count.
+    protected abstract void accumulateRowSizes(int[] positionOffsets, int positionCount, int[] rowSizes);
+
+    public void setupDecodedBlocksAndPositions(DecodedBlockNode decodedBlockNode, int[] positions, int positionCount)
+    {
+        requireNonNull(decodedBlockNode, "decodedBlockNode is null");
+        requireNonNull(positions, "positions is null");
+
+        this.positions = positions;
+        this.positionCount = positionCount;
+        this.positionsOffset = 0;
+
+        setupDecodedBlockAndMapPositions(decodedBlockNode);
+    }
+
+    protected void setupDecodedBlockAndMapPositions(DecodedBlockNode decodedBlockNode)
+    {
+        requireNonNull(decodedBlockNode, "decodedBlockNode is null");
+        decodedBlock = (Block) mapPositions(decodedBlockNode).getDecodedBlock();
+    }
+
+    protected void resetNullsBuffer()
+    {
+        nullsBufferIndex = 0;
+        remainingNullsCount = 0;
+        hasEncodedNulls = false;
+    }
+
+    protected DecodedBlockNode mapPositions(DecodedBlockNode decodedBlockNode)
```

Comment:
This part of "remapping" logic looks very similar to the "block unpeel" work done in `Columnar{Array, Map, Row}` and `BlockFlattener`. Just curious why we cannot use the "unpeeled block" generated by `BlockFlattener`, for example?

Reply:
@wenleix  hi Wenlei, both BlockFlattener.flatten() and ColumnarXXX.toColumnarXXX() preserve the last Dictionary/Rle wrapping, therefore we still need to do remapping here. This mapPositions() peels off the last Dictionary/Rle wrapping.

Reply:
@yingsu00 : I was originally refer to https://github.com/prestodb/presto/blob/df5a089efaea21ed3d988228ef40605d7fd079e1/presto-spi/src/main/java/com/facebook/presto/spi/block/BlockFlattener.java#L62-L86

But now I see it's different: here it did "position mapping" to the leaf level data block.

Maybe rename the method to `setPositionMapToLeafDataBlock`? And add some comments for this method? (Something like the following, feel free to edit/correct it if I don't understand correctly ) 
```
Setup the position mapping to leaf data block. Consider the following dictionary block
leaf data block: ['a', 'b', 'c']
ids: [0, 1, 2, 0, 2, 1, 2, 2, 2]

The data it represents is ['a', 'b', 'c', 'a', 'c', 'b', 'c', 'c', 'c']. mappedPositions will be XXXX.

Now suppose we want to append positions 0, 2, 5, 7 in this Dictionary Block to the buffer, and we will end up appending mappedPositions[0], mappedPositions[2], mappedPositions[5], mappedPositions[7] in the leaf data block to the buffer, which is X, X, X, X.
```

Reply:
Thanks for the clarification. How about the following name and comment?

```
    /**
     * Map the positions for DictionaryBlock and RunLengthEncodedBlock to the corresponding positions in the leaf block.
     * For example, positions [0, 2, 5] on DictionaryBlock with leaf block ['a', 'b', 'c'] and ids [1, 1, 2, 0, 2, 1]
     * will be mapped to [1, 2, 1], and the copied data will be ['b', 'c', 'b'].
     */
    protected DecodedBlockNode mapPositionsToLeafBlock(DecodedBlockNode decodedBlockNode)
```